src/lib.rs

```diff
@@ -1295,30 +1252,19 @@ pub unsafe trait TryFromBytes {
     where
         Self: KnownLayout,
     {
-        let maybe_self = Ref::<_, Self>::new_known_layout_name_to_be_bikeshedded(bytes)?;
+        let candidate = Ptr::from_ref(bytes).try_cast_into_no_leftover::<Self>()?;
+
+        // SAFETY: `candidate` has no uninitialized sub-ranges because it
+        // derived from `bytes: &[u8]`, and is therefore at least as-initialized
+        // as `Self`.
+        let candidate = unsafe { candidate.assume_as_initialized() };
 
-        // SAFETY:
-        // - Since `bytes` is an immutable reference, we know that no mutable
-        //   references exist to this memory region.
-        // - Since `[u8]` contains no `UnsafeCell`s, we know there are no
-        //   `&UnsafeCell` references to this memory region.
-        // - Since `TryFromBytes` types may not contain any `UnsafeCell`s, the
-        //   requirement that all references contain `UnsafeCell`s at the same
-        //   offsets is trivially satisfied.
-        // - All bytes of `bytes` are initialized.
-        //
         // This call may panic. If that happens, it doesn't cause any soundness
         // issues, as we have not generated any invalid state which we need to
         // fix before returning.
-        if unsafe { !Self::is_bit_valid(maybe_self.as_ptr()) } {
-            return None;
-        }
+        let candidate = candidate.check_valid();
 
-        // SAFETY:
-        // - `is_bit_valid` guarantees that `*maybe_self` contains a valid
-        //   `Self`.
-        // - `TryFromBytes` types may not contain any `UnsafeCell`s.
-        Some(unsafe { maybe_self.into_ref_unchecked() })
+        candidate.map(MaybeAligned::as_ref)
```

Comment:
Can you coalesce these?

Reply:
Discussed offline. Decided against doing so, because the panic comment is specific to `check_valid`.